Run <pre> contents through highlighting tool

sideshowbarker opened the pull request:
This change adds the mechanism necessary for running <pre> contents through
an external tool. Once we actually have a syntax-highlighting tool
available we can pass the <pre> contents to, we can update this change to
use the real tool.

However, since we don’t yet have that tool, then at this point we just use
/bin/cat and run the HTML contents of all <pre> elements through it and
then just write that HTML back to the spec output. So it’s not adding any
actual syntax-highlighting markup yet.

Even though this isn’t really changing the <pre> contents at all, even the
mechanism of just running the contents through /bin/cat more than doubles
the wattsi execution time.

In my environment before this change, the wattsi execution time was about 6
seconds. But after this change, the execution time is about 14 seconds. So
this change on its own adds about 8 seconds.

Part of the reason behind that big increase is, for each <pre> this
mechanism actually runs each <pre> element through the external tool three
times: Once while generating the single-page spec output, once more while
generating the split/multipage output, and one last time while generating
the dev-version output (though for that it skips all the IDL <pre>s,
because those don’t get output in the dev edition).

Addresses https://github.com/whatwg/html-build/issues/113

sideshowbarker commented:
This is ready for review

domenic commented on src/wattsi.pas:
```diff
@@ -1817,14 +1932,60 @@ procedure Save(const Document: TDocument; const FileName: AnsiString; const InSp
+            if (Pos(Element.GetAttribute('class').AsString, 'idl') <> 0) then
+               Highlighter.Parameters.Add('webidl')
+            else
+            if (Pos(Element.GetAttribute('class').AsString, 'css') <> 0) then
```
Should we include js in here as well for when we fix the source document?

sideshowbarker replied:
+            if (Pos(Element.GetAttribute('class').AsString, 'css') <> 0) then

> Should we include js in here as well for when we fix the source document?

Yes, added in 00969db

sideshowbarker commented:
This is ready to land, if anybody’s interested in reviewing. I can squash the commits if that helps.

It can be merged at any time — I mean prior to https://github.com/whatwg/html/pull/3768 and https://github.com/whatwg/html-build/pull/162. Without the https://github.com/whatwg/html-build/pull/162 changes, wattsi after these changes just behaves as always, without attempting to do any syntax highlighting. The addition of the extra command-line argument that https://github.com/whatwg/html-build/pull/162 causes it to be called with is what triggers the highlighting behavior.

sideshowbarker commented:
> New version LGTM too.

Cool — thanks for the review

> Only comment is that I think some variables could be clearer. E.g., if I am understanding correctly:
>
> AsJSON -> CurrentlyInHighlightedElement
> JSONContents -> CurrentHighlightedElementJSON
> HTMLContents -> CurrentHighlightedElementFallbackHTML
> WriteJSON -> WriteJSONIfInHighlightedElement

100% agreed — those are much clearer, so I’ve changed them all to the above

> InPreElement: seems to be unused

Yep, thanks for catching that — I’ve removed it
